### PR TITLE
MumbleBee, a bee that connects to mumble servers

### PR DESCRIPTION
**Pull request by momo-aux**
… to channels

on a mumble Server.

### CODE REVIEW

File: bees/mumblebee/mumblebee.go [hunk 1]
```diff
@@ -0,0 +1,250 @@
...
+	}
+	//*/
+
+	config.Attach(gumbleutil.AutoBitrate)
+
+	connecting := false
+	disconnected := true
+	waitForDisconnect := false
+	for {
+		// loop on IRC connection events
```
Comment (muesli) [inline bees/mumblebee/mumblebee.go]: Heh, classic copy/paste error.

File: bees/mumblebee/mumblebee.go [hunk 2]
```diff
@@ -0,0 +1,250 @@
...
+		action.Options.Bind("user", &user)
+		action.Options.Bind("text", &text)
+
+		mod.client.Users.Find(user).Send(text)
+
+	case "send_to_channel":
+
+		channel, text := "", ""
+
+		action.Options.Bind("channel", &channel)
```
Comment (muesli) [inline bees/mumblebee/mumblebee.go]: ```channel``` doesn't seem to be used here. What's the difference to the "say" action below?

File: bees/mumblebee/mumblebee.go [hunk 3]
```diff
@@ -0,0 +1,250 @@
...
+				mod.Logln("Connected to:", mod.server)
+				if err != nil {
+					mod.Logln("Failed to connect to Mumble:", mod.server, err)
+					connecting = false
+				}
+			}
+		}
+		select {
+		case status := <-mod.connectedState:
+			mod.Logln("status case")
```
Comment (muesli) [inline bees/mumblebee/mumblebee.go]: Probably best to remove this and the other ```status xxx``` debug/development log-messages below so we don't spam our user's logs.

File: bees/mumblebee/mumblebeefactory.go [hunk 4]
```diff
@@ -0,0 +1,203 @@
...
+func (factory *MumbleBeeFactory) Options() []bees.BeeOptionDescriptor {
+	opts := []bees.BeeOptionDescriptor{
+		{
+			Name:        "address",
+			Description: "Hostname of Mumble server, eg: de.clanwarz.com:64738",
+			Type:        "string",
+			Mandatory:   true,
+		},
+		{
+			Name:        "channel",
```
Comment (muesli) [inline bees/mumblebee/mumblebeefactory.go]: Can you join multiple channels concurrently on a Mumble server? In this case, we could make it ```channels``` of ```[]string``` type. We could also get rid of the ```move``` action then.

File: bees/mumblebee/mumblebeefactory.go [hunk 5]
```diff
@@ -0,0 +1,203 @@
...
+// Events describes the available events provided by this Bee.
+func (factory *MumbleBeeFactory) Events() []bees.EventDescriptor {
+	events := []bees.EventDescriptor{
+		{
+			Namespace:   factory.Name(),
+			Name:        "message",
+			Description: "A message was received over Mumble",
+			Options: []bees.PlaceholderDescriptor{
+				{
+					Name:        "text",
```
Comment (muesli) [inline bees/mumblebee/mumblebeefactory.go]: We could also emit the channel this message was received on.

File: hives.go [hunk 6]
```diff
@@ -49,4 +49,5 @@ import (
 	_ "github.com/muesli/beehive/bees/twiliobee"
 	_ "github.com/muesli/beehive/bees/twitterbee"
 	_ "github.com/muesli/beehive/bees/webbee"
+	_ "github.com/muesli/beehive/bees/mumblebee"
```
Comment (muesli) [inline hives.go]: Sort them alphabetically. I suggest running ```go fmt *.go``` which does that for you.

File: bees/mumblebee/mumblebee.go [hunk 7]
```diff
@@ -0,0 +1,250 @@
...
+	}
+
+	config.Attach(gumbleutil.Listener{
+		TextMessage: func(e *gumble.TextMessageEvent) {
+			ev := bees.Event{
+				Bee:  mod.Name(),
+				Name: "message",
+				Options: []bees.Placeholder{
+					{
+						Name:  "channels",
```
Comment (muesli) [inline bees/mumblebee/mumblebee.go]: This option isn't defined in the factory. I also assume it should be ```channel``` (singular) of type ```string``` in that case.

File: bees/mumblebee/mumblebee.go [hunk 8]
```diff
@@ -140,7 +135,7 @@ func (mod *MumbleBee) Run(eventChan chan bees.Event) {
 				Name: "message",
 				Options: []bees.Placeholder{
 					{
-						Name:  "channels",
+						Name:  "channel",
 						Type:  "[]string",
```
Comment (muesli) [inline bees/mumblebee/mumblebee.go]: Should be ```string``` then, not an array of strings.

File: bees/mumblebee/mumblebeefactory.go [hunk 9]
```diff
@@ -0,0 +1,203 @@
...
+	}
+	return events
+}
+
+// Actions describes the available actions provided by this Bee.
+func (factory *MumbleBeeFactory) Actions() []bees.ActionDescriptor {
+	actions := []bees.ActionDescriptor{
+		{
+			Namespace:   factory.Name(),
+			Name:        "say",
```
Comment (muesli) [inline bees/mumblebee/mumblebeefactory.go]: This Action can be removed as it doesn't exist any longer.

File: bees/mumblebee/mumblebeefactory.go (hunk 5 above)
Comment (muesli) [inline bees/mumblebee/mumblebeefactory.go]: The ```channel``` is now being emitted, but is still missing here.

File: bees/mumblebee/mumblebee.go [hunk 10]
```diff
@@ -0,0 +1,240 @@
...
+	config.Attach(gumbleutil.Listener{
+		TextMessage: func(e *gumble.TextMessageEvent) {
+			ev := bees.Event{
+				Bee:  mod.Name(),
+				Name: "message",
+				Options: []bees.Placeholder{
+					{
+						Name:  "channel",
+						Type:  "[]string",
+						Value: e.Channels,
```
Comment (muesli) [inline bees/mumblebee/mumblebee.go]: I'm confused. ```e.Channels``` seems to indicate a message can be received on multiple channels? Isn't it always just one channel?

File: bees/mumblebee/mumblebeefactory.go [hunk 11]
```diff
@@ -0,0 +1,208 @@
...
+				},
+				{
+					Name:        "user",
+					Description: "The user that sent the message",
+					Type:        "string",
+				},
+				{
+					Name:        "channels",
+					Description: "The channels where the message was received",
+					Type:        "string",
```
Comment (muesli) [inline bees/mumblebee/mumblebeefactory.go]: This sould be a ```[]string```.

File: bees/mumblebee/mumblebeefactory.go [hunk 12]
```diff
@@ -0,0 +1,208 @@
...
+		{
+			Namespace:   factory.Name(),
+			Name:        "send_to_channel",
+			Description: "Sends a message to a channel",
+			Options: []bees.PlaceholderDescriptor{
+				{
+					Name:        "channel",
+					Description: "Which channel to send the message to",
+					Type:        "string",
+					Mandatory:   true,
```
Comment (muesli) [inline bees/mumblebee/mumblebeefactory.go]: This should be false. If no channel gets provided, this bee will send to the current channel.

File: bees/mumblebee/mumblebeefactory.go [hunk 13]
```diff
@@ -0,0 +1,208 @@
...
+					Name:        "channel",
+					Description: "The channel to move in",
+					Type:        "string",
+					Mandatory:   true,
+				},
+			},
+		},
+		{
+			Namespace:   factory.Name(),
+			Name:        "play",
```
Comment (muesli) [inline bees/mumblebee/mumblebeefactory.go]: This action doesn't actually exist yet.

**Comment by muesli** [comment]
Thank you @momo-aux!

**Comment by Joshfindit** [comment]
P.s. Excellent move from the previous title to MumbleBee. Much clearer. (plus, I think this is a great name in general)

**Comment by muesli** [comment]
@Joshfindit I'm not quite sure I understand what you're referring to?

**Comment by Joshfindit** [comment]
@muesli:

`Bee that can react on text messages, send text messages and move…` = Confusing title. Sounds like SMS.
`MumbleBee, a bee that connects to mumble servers` = Much clearer.
`MumbleBee` = Catchy name.

**Comment by muesli** [comment]
Understood & agreed now :)

**Pull request merged**